client/client-lib/src/lib.rs

```diff
@@ -341,13 +341,19 @@ impl<T: AsRef<ClientConfig> + Clone> Client<T> {
         self.submit_tx_with_change(tx, &mut rng).await
     }
 
-    async fn submit_tx_with_change<R: RngCore + CryptoRng>(
+    /// Submits a transaction to the fed, making change using our change module
+    ///
+    /// If the submission fails, no changes to the DB will occur.
+    pub async fn submit_tx_with_change<R: RngCore + CryptoRng>(
         &self,
         tx: TransactionBuilder,
         rng: R,
     ) -> Result<TransactionId> {
-        let final_tx = tx.build(self, rng).await;
-        Ok(self.context.api.submit_transaction(final_tx).await?)
+        let mut dbtx = self.context.db.begin_transaction().await;
+        let final_tx = tx.build(self, &mut dbtx, rng).await;
+        let result = self.context.api.submit_transaction(final_tx).await?;
```

Comment:
This has an opposite problem. submit can succeed, while commit might fail / never happen.

Reply:
Note: Even though not "correct", this change is probably better than what we have before. So I'm not saying that we shouldn't land it, just that it's not exactly right approach either.